### PR TITLE
Add Task as a builtin module/type

### PR DESCRIPTION
**Pull request by lukewilliamboswell**
This is a twin PR for https://github.com/roc-lang/roc/pull/6836 in the roc-lang/roc repo. 

It updates the FFI in this platform to return exclusively Tasks, as well as removes this platform's Task module (and InternalTask).

### CODE REVIEW

**Comment by lukewilliamboswell** [comment]
Rebased on main, and updated to use latest roc-land/roc branch. 

Currently fails with the following;

```
$ ROC=roc EXAMPLES_DIR=examples/ bash ci/all_tests.sh
<------SNIP------>
++ uname -m
+ architecture=arm64
+ for roc_file in '$EXAMPLES_DIR*.roc'
+ roc build --prebuilt-platform --linker=legacy examples/command.roc
0 errors and 0 warnings found in 260 ms
 while successfully building:

    examples/command
+ for roc_file in '$EXAMPLES_DIR*.roc'
+ roc build --prebuilt-platform --linker=legacy examples/dir.roc
thread 'main' panicked at /Users/luke/.cargo/git/checkouts/inkwell-55729a4c43239568/d1a5963/src/values/enums.rs:302:13:
Found StructValue(StructValue { struct_value: Value { name: "load_result", address: 0x600000866900, is_const: false, is_null: false, is_undef: false, llvm_value: "  %load_result = load { [0 x i64], [3 x i64], i8, [7 x i8] }, ptr %result_value, align 8, !dbg !10", llvm_type: "{ [0 x i64], [3 x i64], i8, [7 x i8] }" } }) but expected PointerValue variant
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by Anton-4** [comment]
I apologize for the many commits, I'm currently on Ubuntu and it's easiest to make all the changes using the github website.

**Comment by Anton-4** [comment]
Right, backpassing is deprecated now, need to update some code...

**Comment by Anton-4** [comment]
I'm going to try to make an llm perform all the backpassing changes :p

File: crates/roc_host/src/roc.rs [hunk 1]
```diff
@@ -614,15 +614,15 @@ fn to_roc_write_error(err: std::io::Error) -> roc_app::WriteErr {
         // TODO investigate support the following IO errors may need to update API
         ErrorKind::WriteZero => roc_app::WriteErr::WriteZero(),
         // TODO investigate support the following IO errors
-        // std::io::ErrorKind::FileTooLarge <- unstable language feature
```
Comment (smores56) [inline crates/roc_host/src/roc.rs:617]: These are false positives in your backpassing removal attempt, methinks.

File: examples/command.roc [hunk 2]
```diff
@@ -13,25 +12,31 @@ respond : Request, Model -> Task Response [ServerErr Str]_
...
-                                    - Exceeding resource limits
-                                    - System shutdown
-                                    - Parent process terminating child processes
-                                    - ...
-                                    """)
+            Task.err
+                (
+                    ServerErr
+                        """
+                        Command was killed by signal. This can happen for everal reasons:
```
Comment (smores56) [inline examples/command.roc:31]: typo in existing code: everal -> several

File: examples/todos.db
Comment (smores56) [inline examples/todos.db:1]: Should this file be here? I expect it should be added to the `.gitignore`

File: platform/Env.roc [hunk 3]
```diff
@@ -76,27 +67,20 @@ var = \name ->
...
-                result
-                |> Result.mapErr (\{} -> VarNotFound)
-                |> Result.try
-                    (\varStr ->
-                        Decode.fromBytes (Str.toUtf8 varStr) (EnvDecoding.format {})
-                        |> Result.mapErr (\_ -> DecodeErr TooShort)))
-    |> InternalTask.fromEffect
+    PlatformTask.envVar name
+    |> Task.mapErr \{} -> VarNotFound
+    |> Task.await \varStr ->
```
Comment (smores56) [inline platform/Env.roc:72]: nitpick: maybe this should be split into a separate definition?

File: examples/todos.db
Comment (Anton-4) [reply examples/todos.db:1]: The example connects to an existing database, this seems better than creating a new one every time. That would risk writing over an existing database if someone tries to adapt the example for their use case. An other approach; throwing an error if a .db already exists, would make testing more annoying.

**Comment by lukewilliamboswell** [comment]
@smores56 @bhansconnect 

Updated the platform to use `{}` instead of `[]` across the host boundary and ensure we use a `RocResult` too.

Unfortunately I am seeing the below issue which means I can get it to build. I can investigate more sometime Monday, but just wanted to give you an update in case you have time to look at this before me. 

```
$ roc build platform/libapp.roc
🔨 Rebuilding platform...
thread 'main' panicked at crates/compiler/gen_llvm/src/llvm/build.rs:3391:30:
called `Result::unwrap()` on an `Err` value: AlignmentError("The src_align_bytes argument to build_memcpy was not a power of 2.")
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Review by smores56 — APPROVED** [review]

File: examples/todos.roc [hunk 4]
```diff
@@ -143,32 +149,14 @@ isSqliteInstalled =
...
         Err _ -> Task.err Sqlite3NotInstalled
 
-logRequest : Request -> Task {} *
+logRequest : Request -> Task {} [StdoutErr Str]
 logRequest = \req ->
     datetime = Utc.now! |> Utc.toIso8601Str
 
-    Stdout.line! "$(datetime) $(Http.methodToStr req.method) $(req.url)"
+    Stdout.line "$(datetime) $(Http.methodToStr req.method) $(req.url)"
+    |> Task.mapErr \err -> StdoutErr (Inspect.toStr err)
```
Comment (smores56) [inline examples/todos.roc:157]: Maybe we should keep this as a `StdoutErr Stdout.Err` instead, to be compatible with other `Stdout.Err` handling code.

File: platform/Http.roc [hunk 5]
```diff
@@ -129,31 +127,33 @@ send = \req ->
...
-    |> Task.mapErr HttpErr
+    { variant, body, metadata } =
+        PlatformTask.sendRequest (Box.box reqToHost)
+            |> Task.mapErr! \_ -> crash "unreachable"
+
+    when variant is
+        "Timeout" -> Task.err (HttpErr (Timeout timeoutMilliseconds))
+        "NetworkErr" -> Task.err (HttpErr NetworkError)
+        "BadStatus" ->
+            Task.err
```
Comment (smores56) [inline platform/Http.roc:138]: nitpick: this is a good candidate for pipelining for readability IMO

**Review by smores56 — APPROVED** [review]

File: examples/todos.roc (hunk 4 above)
Comment (smores56) [reply examples/todos.roc:157]: It's an example, nbd

File: platform/Http.roc (hunk 5 above)
Comment (smores56) [reply platform/Http.roc]: Fixed!

**Review by smores56 — APPROVED** [review]

**Review by Anton-4 — DISMISSED** [review]

**Comment by smores56** [comment]
I'll clean this up now!

**Comment by smores56** [comment]
Not much was really needed, we are now just waiting for a `basic-cli` v0.15 release to update our scripts with, then we can merge this

**Review by smores56 — DISMISSED** [review]

**Review by Anton-4 — APPROVED** [review]

**Pull request merged**